crates/red_knot_workspace/src/site_packages.rs

```diff
@@ -25,26 +77,92 @@ use ruff_db::system::{System, SystemPath, SystemPathBuf};
 /// will be `/opt/homebrew`, and `site-packages` will be at
 /// `/opt/homebrew/lib/python3.X/site-packages`.
 ///
-/// This routine does not verify that `sys_prefix_path` points
-/// to an existing directory on disk; it is assumed that this has already
-/// been checked.
-///
 /// [`sys.prefix`]: https://docs.python.org/3/library/sys.html#sys.prefix
+#[derive(Debug, PartialEq, Eq)]
+struct SysPrefixPath(SystemPathBuf);
+
+impl SysPrefixPath {
+    fn new(unvalidated_path: impl AsRef<SystemPath>, system: &dyn System) -> Option<Self> {
+        let unvalidated_path = unvalidated_path.as_ref();
+        let canonicalized = system
+            .canonicalize_path(unvalidated_path)
+            .unwrap_or_else(|_| unvalidated_path.to_path_buf());
+        system
+            .is_directory(&canonicalized)
+            .then_some(Self(canonicalized))
+    }
+
+    fn from_executable_home_path(path: &PythonHomePath, system: &dyn System) -> Option<Self> {
+        let candidate = path.parent()?;
+        system
+            .is_directory(candidate)
+            .then(|| Self(candidate.to_path_buf()))
+    }
+}
+
+impl Deref for SysPrefixPath {
+    type Target = SystemPath;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl fmt::Display for SysPrefixPath {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "`sys.prefix` path {}", self.0)
+    }
+}
+
+/// E.g. `12` for Python 3.12.4
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+struct PythonMinorVersion(u8);
+
+impl PythonMinorVersion {
+    fn from_version_string(version_string: &str) -> Option<Self> {
+        version_string.parse().ok().map(Self)
+    }
+
+    const fn free_threaded_build_available(self) -> bool {
+        self.0 >= 13
+    }
```

Comment:
Nit: Should we implement this as a method on `TargetVersion`?

Reply:
I don't think we should use the `TargetVersion` type here. The user might be using a Python venv constructed creating a Python version we don't support. In theory that's fine as long as they pass `--target-version` with a Python version we _do_ support

Reply:
I don't think unsupported python versions are a concern because constructing the Program settings would fail to construct and red knot aborts immediately (but we may want to design TargetVersion in a way so that this isn't a problem)

Reply:
> constructing the Program settings would fail to construct

Why? Again, the Python version that the user is running locally could be a completely different Python version to the one they've specified with `--target-version`. The former is what's relevant here, not the latter.